app/models/schedule_intent.rb

```diff
@@ -0,0 +1,12 @@
+class ScheduleIntent < PublishIntent
```

Comment:
What is `ScheduleIntent` used for? Why does it inherit from `PublishIntent`?

Can you separate out creating the `ScheduleIntent` class and it's associated commits into a separate commit and explain what it's for?

Reply:
ScheduleIntent no longer inherits from PublishIntent, since we are not giving the option to specify a change note.

The controller was getting big and it’s an attempt to make it simpler. This class follows the pattern of the PublishIntent model. As Chris said ‘it’s a self-contained model for the schedule-specific changenote and validation logic, and we also need a hook for displaying validation errors in the frontend.’
Having it a separate class prevents from complicating PublishIntent and I’m not sure how we could validate that the change_note is present for immediate publish but not for scheduled there. Plus we don’t have to make changes to StepNavPresenter. 
We will make it a separate commit.

Reply:
Have now pulled it out into a separate commit 👍

Reply:
I'm still not sure what it's being used for as there is also validation logic in the `StepByStepPages` model.

It feels like things are getting over-complicated.
I thought that it would just be possible to:
1. Add a new button on publish-and-delete to direct users to the schedule screen
2. Add a new route, controller action and screen to capture the `scheduled_at` date
3. validate that the date the user supplied is in the future
4. Save `scheduled_at`
5. Call `StepNavPublisher.schedule_for_publishing` to schedule the job.

I'm not clear where `ScheduleIntent` fits into that.

Reply:
What would you suggest doing as an alternative? Aga & I have talked it through and we think this is nicely encapsulated - the alternative would be complicating the controller. If anything I'd suggest we take the duck typing even further and encapsulate the bulk of the `publish`/`schedule` controller methods inside the intents, e.g.

step_by_step_pages_controller.rb:

```ruby
def publish
  set_current_page_as_step_by_step
  @publish_intent.action!
end

def schedule
  set_current_page_as_step_by_step
  @schedule_intent.action!
end

```

publish_intent.rb:

```ruby
def action!
  if request.post?
    @publish_intent = PublishIntent.new(params)
    if @publish_intent.valid?
      publish_page
      generate_internal_change_note
      set_change_note_version
      redirect_to @step_by_step_page, notice: "'#{@step_by_step_page.title}' has been published."
    end
  end
end
```

schedule_intent.rb:

```ruby
def action!
  if request.post?
    @schedule_intent = ScheduleIntent.new(params)
    if @schedule_intent.valid?
      @step_by_step_page.update_attribute(:scheduled_at, params[:scheduled_at])
      schedule_to_publish
      generate_internal_change_note
      set_change_note_version
      redirect_to @step_by_step_page, notice: "'#{@step_by_step_page.title}' has been scheduled to publish."
    end
  end
end
```

Reply:
The default update type is "minor", so you don't need to specifically set that anywhere, which is what I think you are trying to do with `ScheduleIntent`? So the code can be kept quite simple.

in `step_by_step_pages_controller.rb`:

```
  def schedule
    set_current_page_as_step_by_step
    if request.post?
        if @step_by_step_page.update(:scheduled_at, params[:scheduled_at])
          StepNavPublisher.schedule_for_publishing(@step_by_step_page)
          generate_internal_change_note
          set_change_note_version
          redirect_to @step_by_step_page, notice: "'#{@step_by_step_page.title}' has been scheduled."
      end
    end
  end
```

`@step_by_step_page.update(:scheduled_at, params[:scheduled_at])` will run the validations in the model, including `in_future`. Any validation errors should be stored in `@step_by_step_page.errors`

So to show the error in the view, you could have:

```
<% if @step_by_step_page.errors.any? %>
  <%= render "shared/steps/form_errors", resource: @step_by_step_page %>
<% end %>
```

Reply:
For neatness, you could encapsulate `StepNavPublisher.schedule_for_publishing(@step_by_step_page)` in a method.

`PublishIntent` is a model, so it doesn't feel right to add controller actions in there.

Reply:
Have now removed ScheduleIntent 👍